llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -914,3 +919,64 @@ bool SpecConstantsPass::collectSpecConstantDefaultValuesMetadata(
 
   return true;
 }
+
+void replaceSpecConstsWithDefaultValue(const Module &M,
+                                       const std::vector<CallInst *> &CIs) {
+  const DataLayout &DL = M.getDataLayout();
+  for (CallInst *CI : CIs) {
+    Value *Dst = CI->getArgOperand(0);
+    Value *Src = CI->getArgOperand(2);
+    Align DstAlign = DL.getABITypeAlign(Dst->getType());
+    Align SrcAlign = DL.getABITypeAlign(Src->getType());
+    Type *StructureType =
+        Src->stripPointerCasts()->getType()->getContainedType(0);
+    uint64_t SpecConstSize = DL.getTypeAllocSize(StructureType);
+
+    IRBuilder B(CI);
+    B.CreateMemCpy(Dst, DstAlign, Src, SrcAlign, SpecConstSize);
+    CI->removeFromParent();
+    CI->deleteValue();
+  }
+}
+
+static bool checkModuleContainsSpecConsts(const Module &M) {
+  for (const Function &F : M.functions()) {
+    if (F.getName().startswith(SYCL_GET_SCALAR_2020_SPEC_CONST_VAL) ||
+        F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL))
+      return true;
+  }
+
+  return false;
+}
+
+/// Function generates a copy of Module and replaces all spec constant uses
+/// memcpy intrinsic.
+/// If the module doesn't contain spec constants then nullptr is returned.
+std::unique_ptr<Module>
+llvm::generateDeviceImageWithDefaultSpecConstants(const Module &M) {
+  if (!checkModuleContainsSpecConsts(M))
+    return nullptr;
+
+  std::unique_ptr<Module> M2 = CloneModule(M);
```

Comment:
Cloning modules should not be a part of a pass. The pass handles spec constants in a provided module. If we need to create several modules, that should be done outside of the pass scope.

Reply:
Yes, I understand, but I still think that it would be better to have cloning logic outside of `SpecConstants.cpp`, because strictly speaking, it is not related to `__syclGetSpecializationConstant*` intrinsics handling